### PR TITLE
feat: add sql component

### PR DESCRIPTION
**Pull request by zidanehakim**
Because

- We need to integrate sql into our data pipeline.
- This will allow us to cover data movement user stories

This commit

- Allow client to connect to their sql database through VDP
- Create execution function to handle following SQL tasks

- [x] TASK_INSERT 
- [x] TASK_UPDATE
- [x] TASK_SELECT
- [x] TASK_DELETE
- [x] TASK_CREATE_TABLE
- [x] TASK_DROP_TABLE

### CODE REVIEW

**Comment by zidanehakim** [comment]
I have opened a new PR regarding the AWS RDS component. I had GPG key and component test issues with the previous PR (#191 ). This PR will be the final one. Thank you!

File: data/awsrds/v0/main.go [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+	var err error
+
+	// check if the password field is present, if it is, create a client with username and password else create a client with IAM
+	if e.Setup.GetFields()["password"] != nil {
+		client, err = newClient(e)
+	} else if e.Setup.GetFields()["password"] == nil && e.Setup.GetFields()["aws-access-key-id"] != nil && e.Setup.GetFields()["aws-secret-access-key"] != nil {
+		client, err = newClientIAM(e)
+	} else {
+		return nil, fmt.Errorf("some fields are missing, please fill in the required fields")
+	}
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:203]: Question
```
	// check if the password field is present, if it is, create a client with username and password else create a client with IAM
	if e.Setup.GetFields()["password"] != nil {
		client, err = newClient(e)
	} else if e.Setup.GetFields()["password"] == nil && e.Setup.GetFields()["aws-access-key-id"] != nil && e.Setup.GetFields()["aws-secret-access-key"] != nil {
		client, err = newClientIAM(e)
	} else {
		return nil, fmt.Errorf("some fields are missing, please fill in the required fields")
	}
```

Should this logic include to newClient?
e.g.
newClient: check if the users input a password
  - if yes -> newTraditionalClient (Probably need to tune the name)
  - if no -> newIAMClient (Probably need to tune the name)

File: data/awsrds/v0/main.go [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+		return nil, fmt.Errorf("error creating SQL client: %v", err)
+	}
+	defer client.Close()
+
+	for _, input := range inputs {
+		var output *structpb.Struct
+
+		switch e.Task {
+		case taskUpsert:
+			tableName := input.GetFields()["table-name"].GetStringValue()
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:215]: Important
I think making dependency clean is important.

Could you take a look [this comment](https://github.com/instill-ai/component/pull/187#discussion_r1662946437)?

TLDR
I want to make protobuf struct to component struct.
In our other services, we depend on component struct rather than protobuf struct.

File: data/awsrds/v0/main.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+				err = UpsertUser(client, &tableName, data)
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			output = &structpb.Struct{Fields: map[string]*structpb.Value{
+				"status": {Kind: &structpb.Value_StringValue{StringValue: "Successfully executed SQL statement"}}}}
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:237]:
```
output = &structpb.Struct{Fields: map[string]*structpb.Value{
				"status": {Kind: &structpb.Value_StringValue{StringValue: "Successfully executed SQL statement"}}}}
```

Same as comment above.
If you use the component struct, you can have the benefit of making code cleaner.

[Example PR](https://github.com/instill-ai/component/pull/190/files#diff-e489ebc6b09b1110ffef7a388d25d6199476e1ec1cfc19c60db73e39735bd98dL205)

File: data/awsrds/v0/main.go [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+			return nil, fmt.Errorf("task %s not supported", e.Task)
+		}
+		outputs = append(outputs, output)
+	}
+	return outputs, nil
+}
+
+func UpsertUser(client *sqlx.DB, tableName *string, data *structpb.Struct) error {
+	// Begin constructing SQL statement
+	sqlStatement := "INSERT INTO " + *tableName + " ("
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:288]: Question
I happened to realise that the different database could need different command.
Could you help us check if the commands of the db you support here, postgres, oracle, ..., are same?

File: data/awsrds/v0/main.go [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+
+	// Prepare and execute the statement
+	rows, err := client.Queryx(sqlStatement)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Prepare the result ListValue
+	result := &structpb.ListValue{}
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:358]: Same as comment above.
If you use the component struct, you can have the benefit of making code cleaner.

[Example PR](https://github.com/instill-ai/component/pull/190/files#diff-e489ebc6b09b1110ffef7a388d25d6199476e1ec1cfc19c60db73e39735bd98dL205)

File: data/awsrds/v0/main.go [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+
+	// Check for errors after iterating over rows
+	if err = rows.Err(); err != nil {
+		return nil, fmt.Errorf("error occurred during row iteration: %v", err)
+	}
+
+	return result, nil
+}
+
+func DeleteUser(client *sqlx.DB, tableName *string, criteria *structpb.Struct) error {
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:401]: Important
It seems to delete records rather than a record.
So, please clarify this point in the document.

File: data/awsrds/v0/main.go [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func QueryUser(client *sqlx.DB, query string) error {
+	// Prepare and execute the statement
```
Comment (chuang8511) [inline data/awsrds/v0/main.go:425]: Important
We will not use Query. So, please take it out.

File: data/awsrds/v0/component_test.go [hunk 8]
```diff
@@ -0,0 +1,364 @@
+package awsrds
```
Comment (chuang8511) [inline data/awsrds/v0/component_test.go:1]: Using mock is good!
I quickly go through it now, and I will check deeper after you finish the first modification.

**Review by chuang8511 — COMMENTED** [review]
I finished my first review.
I am not super familiar with SQL.
Please bring up some discussion points if I miss something.

File: data/awsrds/v0/config/tasks.json [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+      "title": "Output",
+      "type": "object"
+    }
+  },
+  "TASK_DELETE": {
+    "instillShortDescription": "Perform a delete operation to update existing records based on specified criteria",
+    "input": {
+      "instillUIOrder": 0,
+      "properties": {
+        "criteria": {
```
[the remark this answers is not in the document]
Comment (zidanehakim) [reply data/awsrds/v0/config/tasks.json:183]: @chuang8511 

The task will perform a delete operation for multiple data entries that fit the criteria (it could also be singular if the criteria are specific enough). I believe deleting singular data won't be a useful feature since it will be covered if the criteria specify a single object, especially if the criteria involve a unique column such as an ID. Also, if it's only one object, I do feel like it's better to do it on their own rather than using vdp. What do you think? Is there any specific cases? If not, I will proceed with specifying it.

File: data/awsrds/v0/main.go [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+	"postgresql://%s:%s@%s/%s", "sqlserver://%s:%s@%s%s", "oracle://%s:%s@%s/%s", "%s:%s@tcp(%s)/%s",
+}
+
+// newClient creates a new SQL client with traditional username and password
+func newClient(e *execution) (*sqlx.DB, error) {
+	cfg := LoadConfig(e)
+
+	DBEndpoint := fmt.Sprintf("%v:%v", cfg.DBHost, cfg.DBPort)
+
+	// Test every engines to find the correct one
```
[the remark this answers is not in the document]
Comment (zidanehakim) [reply data/awsrds/v0/main.go:105]: @chuang8511 

At first, I was thinking of letting the user do so, yet I feel like the component UI is already vertically long enough due to the setup connection. I noticed that every DB being created in AWS RDS has a unique sort of string in its endpoint, making it distinct from one another. The purpose of this design is to:

- Reduce the component UI's height
- Provide convenience for the user since it serves as an auto-detect system

Please let me know your thoughts. I'm personally not sure whether this design will fit into the current development.

[the remark this answers is not in the document]
Comment (chuang8511) [reply data/awsrds/v0/main.go:105]: @zidanehakim 
> Reduce the component UI's height

It is a problem that we should adjust from UI design & frontend in my opinion. 
Some other components also contain this situation.
Thanks for considering that! It is good to have this perspective!

> Provide convenience for the user since it serves as an auto-detect system

auto-detect is good. But, I am not so sure if the users contain 2 sql server with the same setting.
How about we let users choose "postgres", "mysql", "oracle", ..., and we set the raw string for them?
Will it be better?

File: data/awsrds/v0/main.go (hunk 4 above)
Comment (zidanehakim) [reply data/awsrds/v0/main.go:288]: @chuang8511 

Since all of the mentioned databases use SQL, they should have the same, if not very similar, commands. The differences lie in how they manage the data within the databases themselves. Although there are some differences in specific commands for each engine, basic and simple tasks like SELECT, DELETE, and UPSERT are mostly the same.

[https://aws.amazon.com/compare/the-difference-between-mysql-vs-postgresql/#:~:text=In%20a%20MySQL%20database%2C%20you,complex%20SQL%20statements%20using%20functions.](url)

File: data/awsrds/v0/main.go (hunk 1 above)
Comment (zidanehakim) [reply data/awsrds/v0/main.go:203]: @chuang8511 

You mean to put the conditional logic within newClient and make two functions based on the authentication method the user chooses? That would also work.

File: data/awsrds/v0/config/tasks.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (chuang8511) [reply data/awsrds/v0/config/tasks.json:183]: @zidanehakim 
Ah, I just want our document to clarify this delete is for multiple objects if it could do deletion for multiple objects.
I think deleting multiple objects is fine for my imagination of the use case.

File: data/awsrds/v0/main.go (hunk 1 above)
Comment (chuang8511) [reply data/awsrds/v0/main.go:203]: > conditional logic within newClient and make two functions based on the authentication method the user chooses?

Yes! but actually it is only some preference to make code explainable.

File: data/awsrds/v0/main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zidanehakim) [reply data/awsrds/v0/main.go:105]: @chuang8511 

If the user has exactly the same setting for 2 different engines, that would be just on master account, port, and region. However, the host would be always different since it has that unique string kind of id.

For example, this is my host
"database.c9y80ugco7tr.ap-southeast-2.rds.amazonaws.com",

The 'c9y80ugco7tr' is the unique string i've mentioned before

[the remark this answers is not in the document]
Comment (chuang8511) [reply data/awsrds/v0/main.go:105]: Ah! I see! Thanks for your explanation!

Another question
How long does each connection take?
`db, err = sqlx.Open("mysql", dsn)`

Will it influence UX eventually?

[the remark this answers is not in the document]
Comment (zidanehakim) [reply data/awsrds/v0/main.go:105]: @chuang8511 

After performing end-to-end tests on the component, I found that each task took around 2 seconds at most. I believe the connection is quite fast and won't significantly impact UX. Note that my database is small, with only 2 tables and around 20 entries each, so I dont know if it scales larger.

[the remark this answers is not in the document]
Comment (zidanehakim) [reply data/awsrds/v0/main.go:105]: @chuang8511 

I would like to ask for your suggestion. For this component, is it necessary to have the data as an array of JSON inputs to insert multiple entries without using an iterator?

For instance, if I have an array of JSON objects in one line:

css
Copy code
[{"name":"zidane1"},{"name":"zidane2"}]
I can put it in the data field right away as:

css
Copy code
${[{"name":"zidane1"},{"name":"zidane2"}]}
Would this be better? I was thinking that if the user has dozens of data to insert, it would be hard to specify each one individually. Having a one-liner like this would be more convenient.

[the remark this answers is not in the document]
Comment (chuang8511) [reply data/awsrds/v0/main.go:105]: @zidanehakim 
Thanks for testing!

> I found that each task took around 2 seconds at most.

Though the indicator here needs to be carefully defined in the future, we still can easily follow some standards if it does not take too much effort. Process connection for 2 sec is too long in the general standard.

[source of reference](https://blog.sentry.io/whats-the-difference-between-api-latency-and-api-response-time/#:~:text=Just%20to%20have%20a%20number,to%20abandon%20the%20application%2Fwebsite.)
image


>  Note that my database is small, with only 2 tables and around 20 entries each, so I dont know if it scales larger.

Just share, I thought the connection time will be influenced by table & entries count.

[the remark this answers is not in the document]
Comment (chuang8511) [reply data/awsrds/v0/main.go:105]: @zidanehakim 

> I would like to ask for your suggestion. 

I am not quite sure if I understand you correctly. 
Please correct me if you think my response doesn't make sense.

>  I was thinking that if the user has dozens of data to insert

Inserting multiple data in a bulk can reduce the DB connection time, which increase the performance
So, in my imagination, I'd insert data bulk by bulk without using iterator when I build the pipeline.

And, I guess your question is related how a component output referenced by another component input.
I think it is other operator's responsibility, which we (other Instill AI member, probably me or Hui-Tang...) need to take care.

If you think I do not get you, feel free to DM me in Slack. We can quickly huddle to confirm this as well.

**Comment by zidanehakim** [comment]
@chuang8511

I have fixed most of the problems. Please kindly review it for the second round. Thanks!

**Comment by donch1989** [comment]
A quick question: any reason that we choose `sqlx` for the db client?

File: data/awsrds/v0/main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (donch1989) [reply data/awsrds/v0/main.go:103]: We should avoid this type of engine detection. Just because `sql.Open` works doesn't mean it will work for all queries. Even though you've listed only four types of engines here, if there are more, some databases might share the same DSN format but have slightly different commands, which could lead to bugs. Let's make the engine selection explicit so users can choose themselves.

**Comment by zidanehakim** [comment]
@donch1989 

It might just be my preference, both sqxl and sql are actually the same, but I prefer sqlx more since it simpler and make the code cleaner

File: data/awsrds/v0/main.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zidanehakim) [reply data/awsrds/v0/main.go:103]: @donch1989 

Okay, I'll work on adding engine input

**Review by donch1989 — APPROVED** [review]
